### ISSUE TITLE
Running Matlab in Jupyter Notebooks in VS Code (on Windows)

### ISSUE THREAD

**Issue by tylerlekang** [mathworks/jupyter-matlab-proxy#41]
There is currently a VS Code extension for Matlab https://github.com/mathworks/MATLAB-extension-for-vscode/ so I'm not sure if this request/issue should be here or there? But that just allows you to *edit* a .m file fairly nicely, not run it -- in the sense of a notebook with inline outputs for each code sections. (I *strongly* disagree that Live Scripts give an equivalent experience to a notebook. It's not the same, and I don't particularly care for it.)

The ultimate would be to be able to **run** Matlab in a Jupyter Notebook in VS Code, with a nice environment (similar to Python with the Python & Pylance extensions).

Will this be upcoming? Pretty please. :) Thank you!

**Comment by prabhakk-mw** [comment]
Hi @tylerlekang 

VSCode provides multiple ways to run Jupyter notebooks through VSCode.

Currently the `jupyter-matlab-proxy` package depends on the presence of a running Jupyter Notebook Server to relay messages to and from MATLAB.

You can however start a jupyter notebook server, and then connect to it via VSCode and continue notebook development from VSCode.

* Start a Jupyter Lab Session
```bash
jupyter lab 

<snip>
    To access the server, open this file in a browser:
        file:///home/prabhakk/.local/share/jupyter/runtime/jpserver-1012796-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/lab?token=f5c87aa6e805384a82a8774ec366938bc554cd7e8462e1b7
     or http://127.0.0.1:8888/lab?token=f5c87aa6e805384a82a8774ec366938bc554cd7e8462e1b7
<snip>

```

* Click on the link to access Jupyter. _Save this link for future use._
* Start MATLAB via the "MATLAB Kernel" or the "Open MATLAB " buttons
* Carry out the licensing step.
* This will save the licensing options used, and future jupyter servers on this machine will not need the last two steps.

* Now, navigate to VSCode, and open any notebook and click on the "Select Kernel" button
![screenshot_1688619796](https://github.com/mathworks/jupyter-matlab-proxy/assets/64955767/4e476c91-648b-45ce-930f-83bed03f7d73)

* Select existing jupyter server:
![screenshot_1688619814](https://github.com/mathworks/jupyter-matlab-proxy/assets/64955767/695f119f-31e9-4db6-b7eb-1c372f75954b)

* Select MATLAB kernel
![screenshot_1688619672](https://github.com/mathworks/jupyter-matlab-proxy/assets/64955767/07210718-4d27-4624-8c7a-ff91d9312ac0)

* Start using Notebook with MATLAB
![screenshot_1688620273](https://github.com/mathworks/jupyter-matlab-proxy/assets/64955767/67e52c46-a9cc-4caf-a612-0dc1ee51b8bb)

We are working on integrating with VSCode more, and look to enable more modes of access through Jupyter later this year.

Hope this helps!

**Comment by tylerlekang** [comment]
@prabhakk-mw that works perfectly! I did not know that was an option! Maybe in the future there can be a MATLAB for Notebook extension that takes care of even needing to open the JL server in the first place, but that really is a minor inconvenience. The other major thing that is lacking with this way currently is to have function auto-completion and information appear when you type ("IntelliSense"). This is available using the MATLAB extension when editing an .m file (though doesn't display function documentation as well as Python/Pylance extensions).

In fact, using this method addresses all three of the issues I raised for using MATLAB directly in Jupyter Lab

https://github.com/mathworks/jupyter-matlab-proxy/issues/40
https://github.com/mathworks/jupyter-matlab-proxy/issues/42
https://github.com/mathworks/jupyter-matlab-proxy/issues/43

In other words, building the Notebook in VSC instead of JL alleviates all three of these issues. (Not sure how, but none of them happen in VSC Notebook).

**Comment by Will-Zhao0** [comment]
Hi, I'm wondering how you are doing this, where is the buttons?

> * Start MATLAB via the "MATLAB Kernel" or the "Open MATLAB " buttons

**Comment by prabhakk-mw** [comment]
@Will-Zhao0 , I've updated my [answer](https://github.com/mathworks/jupyter-matlab-proxy/issues/41#issuecomment-1623449847) to include a video. 
Hopefully that will help answer the questions you may have.
Thanks

**Comment by tylerlekang** [comment]
For whatever it is worth, what I do in Windows10 is:
- launch "Anaconda prompt"
- activate the environment I want to use (which has jupyterlab and j-ml-proxy installed)
- launch jupyter lab and copy the server address (just like in the video)
- in VSC on a tab for a notebook I want to do Matlab, I select the kernel (again just like in the video)
- first cell I run some kind of version command
- this kicks off the process of starting up Matlab, checking the license, and connecting it to the notebook. I don't have to do anything in the actual JL browser window, and can in fact close it (or just launch the --no-browser version of JL at command line), but I do use it sometimes to kill the kernels if I'm managing multiple

**Comment by prabhakk-mw** [comment]
> For whatever it is worth, what I do in Windows10 is:
> 
> * launch "Anaconda prompt"
> * activate the environment I want to use (which has jupyterlab and j-ml-proxy installed)
> * launch jupyter lab and copy the server address (just like in the video)
> * in VSC on a tab for a notebook I want to do Matlab, I select the kernel (again just like in the video)
> * first cell I run some kind of version command
> * this kicks off the process of starting up Matlab, checking the license, and connecting it to the notebook. I don't have to do anything in the actual JL browser window, and can in fact close it (or just launch the --no-browser version of JL at command line), but I do use it sometimes to kill the kernels if I'm managing multiple

Yes @tylerlekang, you are right, one does not need to actually run anything in the JL interface. 
However, it seemed like a good thing to do when you are setting up for the first time, as this helps confirm that you have a working environment available outside of VSCode.

Additionally, once licensing has been carried out at least once on a machine through the JL interface, this information is cached and is re-used by the MATLAB kernel for subsequent launches.

Licensing cannot be done through the VSCODE interface using the web screens as there isn't a way for us to route those screens into VSCODE.

**Comment by tylerlekang** [comment]
@prabhakk-mw it would be good if somehow it were possible for any external tool (VSC or anything) wanting to do Matlab via Jupyter could connect without having to copy the address every time and re-set the connection. Like if j-ml-proxy had a way to have a permanent ip address/port that would always redirect to the correct/current running jupyter server. I don't know if jupyter allows such a thing to be possible, though!

**Comment by prabhakk-mw** [comment]
> @prabhakk-mw it would be good if somehow it were possible for any external tool (VSC or anything) wanting to do Matlab via Jupyter could connect without having to copy the address every time and re-set the connection. Like if j-ml-proxy had a way to have a permanent ip address/port that would always redirect to the correct/current running Jupyter server. (And then even more ideal, if there were a set of commends you could send to that "middle man layer" that could view and manage the active Jupyter kernels!) I don't know if Jupyter allows such a thing to be possible, though


We are actively working on improving our product for its usage with vscode to be seamless, and get as close as we can in terms of experience to be similar to other kernels. 

Stay tuned. 🙂

**Comment by prabhakk-mw** [comment]
@tylerlekang 
As of v0.16.4 of jupyter-matlab-proxy it should now be possible for you to start a MATLAB notebook without needing to start a Jupyter server separately.

The [install-guide/vscode/README.md](https://github.com/mathworks/jupyter-matlab-proxy/tree/main/install_guides/vscode) has all the details.

Please do let us know of any feedback you may have with using this feature!

Thank you for using the MATLAB Integration for Jupyter.

**Comment by tylerlekang** [comment]
Hello @prabhakk-mw ,

I am not able to run the new command `install-matlab-kernelspec` . Screenshot from Windows Command Prompt:

![Image](https://github.com/user-attachments/assets/5ac213a7-c23b-48f5-b923-61b108a65c80)


I know the readme says to use Python 3.11 at most, but I would not think that would matter for why Cmd does not even recognize the `install-matlab-kernelspec` ?

I also tried closing and opening a new Cmd window. Also tried in Powershell, with no luck.

**Comment by prabhakk-mw** [comment]
@tylerlekang

 it's most likely a PATH issue. 

The install-matlab-kernelspec executable should be located along with the other console scripts like matlab-proxy-app

Can you please verify that you can find them both in the same location and that the location is on PATH? 


tagging, @rashedmyt

**Comment by tylerlekang** [comment]
@prabhakk-mw 

I cannot locate matlab-prox-app either. Can you give me a rough idea of where to look for these? I can find my local Python's `site-packages` folder ... is there a location there I should be looking for? Or else what would be the approximate path I should try searching? I don't see anything like executables in any of these folders.

**Comment by tylerlekang** [comment]
@prabhakk-mw 

I found the path to the script .exe (helpful: doing the uninstall prompted what would be deleted which included the path to the script). I ran the .exe as Administrator and a new Cmd window opened briefly then closed.

If I try to just run the .exe from a Cmd window (opened as Administrator) it has the following error:


I tried opening a new notebook and selecting "Select Another Kernel", but it only shows options for "Python Environments..." (which only shows my only installed regular python kernel on this machine) and "Existing Jupyter Server..."  There is no option for "Jupyter Kernel..." as shown in the Readme

**Comment by rashedmyt** [comment]
@tylerlekang
I was able to reproduce the issue you are facing and was able to determine that the issue is due to the installation of python from Microsoft Store. When using python from other sources (python.org, conda), I did not face this issue.
 
Since you have mentioned that you are not able to use python from other sources, I would suggest you to use virtual environments. There are two ways to do this in your case.
* Using venv
    1. Create a virtual environment in the folder in which you have the Jupyter notebook by running `python -m venv .venv`
    2. Activate the virtual environment by running `.venv/Scripts/activate`
    3. Install jupyer-matlab-proxy in the virtual environment and make sure to run the `install-matlab-kernelspec`.
    4. Note that VS Code only detects virtual environments present in the same or a sub-directory in which the Jupyter Notebook is present when the virtual environments are created using this approach.
* Using pipenv
    1. Install pipenv package by running `python -m pip install pipenv`
    2. Make sure that pipenv is on PATH. The location of the directory where pipenv executable is installed when the above command is run is printed as a warning in yellow text in case the directory is not on PATH.
    3. Create a virtual environment using pipenv by running `pipenv --python 3.12` in a directory.
    4. Activate the virtual environment by running `pipenv shell` in the same directory.
    5. Install jupyter-matlab-proxy in the virtual environment and make sure to run the `install-matlab-kernelspec`.
    6. VS Code should detect the kernels installed in virtual environments created using this approach automatically.

Below is demo video for using venv and getting the kernel working in VS Code. Let us know if you face any further issues

https://github.com/user-attachments/assets/08f8d98e-3d02-4d05-8295-e854599f7d46

**Comment by tylerlekang** [comment]
@rashedmyt 

Outstanding! It is working now, using the venv method. Great! :)
